statedb/statedb.go

```diff
@@ -1,35 +1,741 @@
+/*
+Package statedb contains the implementation of StateDB, a database backed
+structure that holds the state of the blockchain indexed by version (each
+version corresponding to the state at each block).  The StateDB holds a dynamic
+hierarcy of linked merkle trees starting with the mainTree on top, with the
+property that the keys and values of all merkle trees can be cryptographically
+represented by a single hash, the StateDB.Root (which corresponds to the
+mainTree.Root).
+
+Internally all subTrees of the StateDB use the same database (for views) and
+the same transaction (for a block update).  Database prefixes are used to split
+the storage of each subTree while avoiding collisions.  The structure of
+prefixes is detailed here:
+- subTree: `{KindID}{id}/`
+	- arbo.Tree: `t/`
+	- subTrees: `s/`
+		- contains any number of subTree
+	- nostate: `n/`
+	- metadata: `m/`
+		- versions: `v/` (only in mainTree)
+			- currentVersion: `current` -> {currentVersion}
+			- version to root: `{version}` -> `{root}`
+
+Since the mainTree is unique and doesn't have a parent, the prefixes used in
+the mainTree skip the first element of the path (`{KindID}{id}/`).
+- Example:
+	- mainTree arbo.Tree: `t/`
+	- processTree arbo.Tree (a singleton under mainTree): `s/process/t/`
+	- censusTree arbo.Tree (a non-singleton under processTree):
+	`s/process/s/census{pID}/t/` (we are using pID, the processID as the id
+	of the subTree)
+	- voteTree arbo.Tree (a non-singleton under processTree):
+	`s/process/s/vote{pID}/t/` (we are using pID, the processID as the id
+	of the subTree)
+
+Each tree has an associated database that can be accessed via the NoState
+method.  These databases are auxiliary key-values that don't belong to the
+blockchain state, and thus any value in the NoState databases won't be
+reflected in the StateDB.Root.  One of the usecases for the NoState database is
+to store auxiliary mappings used to optimize the capacity usage of merkletrees
+used for zkSNARKS, where the number of levels is of critical matter.  For
+example, we may want to build a census tree of babyjubjub public keys that will
+be used to prove ownership of the public key via a SNARK in order to vote.  If
+we build the merkle tree using the public key as path, we will have an
+unbalanced tree which requires more levels than strictly necessary.  On the
+other hand, if we use a sequential index as path and set the value to the
+public key, we achieve maximum balancing reducing the number of tree levels.
+But then we can't easily query for the existence of a public key in the tree to
+generate a proof.  In such a case, we can store the mapping of public key to
+index in the NoState database.
```

Comment:
Small clarification, that can not easily be queried because would be accessed by index and not bye public key:
```suggestion
But then we can't easily query for the existence of a public key in the tree to
generate a proof, as it requires to know its index.  In such a case, we can
store the mapping of public key to index in the NoState database.
```
(the only addition of the suggestion is `, as it requires to know its index` the rest is diff because of the jump lines)